scripts/shared/lib/deploy_funcs

```diff
@@ -73,6 +73,37 @@ function connectivity_tests() {
     with_context "$target_cluster" remove_resource "${RESOURCES_DIR}/nginx-demo.yaml"
 }
 
+function verify_gw_topology() {
```

Comment:
This is a good thing to test, but as suggested previously please move to a distinct test that would test the different scenarios

Reply:
@mkolesnik 
Sorry, I'm not sure I understand the meaning.
Where exactly verification should be placed and how it will be executed?

Reply:
The tests for scripts are here: https://github.com/submariner-io/shipyard/tree/devel/test/scripts

You would need to add a file to test deploy

Reply:
@mkolesnik 

As per the link you shared, I see that the only place where these scripts are called, are here:
https://github.com/submariner-io/shipyard/blob/devel/.github/workflows/scripts.yml

But I still don't have a full understanding of how exactly it should be placed and how it going to be called.
Appreciate more details.

Reply:
You'll need to create a new test script to test deploy, and call it from that GHA that you mentioned.

Reply:
@mkolesnik 

I looked closely into the github actions and to me it looks like no sense to place it there.
Let me explain my point of view.

I have created the "verify_gw_topology" to ensure that the desired state of the gateway nodes provided in `.shipyard.e2e.yml` or `.shipyard.e2e.ovn.yml` is actually applied.
The importance here is to make sure that if first cluster need to get one gw node and second cluster two gw nodes, that would be the state.

I placed this function near the "verify_gw_status" check function in scripts/shared/deploy.sh as it looked to me right.

If I place the check function within the github actions, there is no real verification of the desired gw state as I can't really set a different configuration per the cluster over there.
I think the live check during the deployment will add a bug value of env validation.

But if you think it should not be there, I can remove it completely.
My main point is that I don't think github action could really check and verify the that kind of flow.

What do you think?

Reply:
You're missing the point entirely..

`verify_gw_status` makes sure theres an established connection before proceeding.
This is because if we don;t wait, there could be a race between the tests starting to run, and connections establishing, and sometimes the tests will entirely fail before the connection established.

You verification is a functional test, testing that when you request more than one gateway you get it.
This is clearly a test, and not any impediment to the testing itself (from shipyard perspective).

Hence, I suggest to put it as a test which tests this functionality, rather than here which doesn't make much sense.

In other words, say I have a function `sum(int, int) int` then I wouldn't put a test each time I call it in the code to test `sum(1,1) == 2`, right?

Reply:
You can create a test which loads a specific settings file. The real part you need to test is `cloud-prepare`  so you run `make clusters` to create them and then `make cloud-prepare` and see that each cluster has the requested number of nodes.

In fact you don't even need a concrete settings file, you can store file contents in your test script in variables, and write them out to a temp file (you can use `mktemp` for creating a temp file).
However if its easier just create the concrete file inside the test directory.

You don't even need multiple clusters. It's enough to create a single cluster with say 2 nodes and after each test unlabel all gateway nodes.

Reply:
Done